### DISCUSSION TITLE
UI UnitManager windows

### DISCUSSION THREAD

**Discussion by bevans2000** [mars-sim/mars-sim#522] [General]
I don't use the Unit windows in the UI that much. But what I have realised is that they are very heavy.
When looking at the Settlement window for Sch. Point it consumes a lot of CPU. Plus the manufacturing panel creates ConcurrentModification exceptions because it is continuously scanning the resources to look for new processes. Also once I close it; it continues to update in the background event though nothing  is showing.
@mokun The only way I can see to resolve this is to go back to using CopyOnWrite maps inside the inventory classes which seems a real shame.
The Building tab in the Settlement window is too large. I think we should have the Building panel as a separate Window and just show a high level summary list in the Settlement window.
Having said all that I think the future is in a Web UI based on a REST api of the simulation engine.

**Comment by mokun** [comment]
> Plus the manufacturing panel creates ConcurrentModification exceptions because it is continuously scanning the resources to look for new processes.

Sure let's optimize that.


> the future is in a Web UI based on a REST api of the simulation engine.

Ok. but in short term, we could go back to using FXGL/JavaFX to build the UI from the ground up and set this UI up for supporting REST API early on. Shall we ?

**Reply by bevans2000** [reply]
> > Plus the manufacturing panel creates ConcurrentModification exceptions because it is continuously scanning the resources to look for new processes.
> 
> Sure let's optimize that.
> 
> > the future is in a Web UI based on a REST api of the simulation engine.
> 
> Ok. but in short term, we could go back to using FXGL/JavaFX to build the UI from the ground up and set this UI up for supporting REST API early on. Shall we ?

Creating a good REST api is a project on it's own but I feel it maybe close. Then if we created one,  why would we not have a web interface? Having a single simulation running online that we shared has to be the ultimate goal.

**Comment by bevans2000** [comment]
Having the AWT Thread trigger the updates is causing the Concurrent exceptions; this is probably indirectly from use of the SwingInvoke later method. I believe generally we expect the UI window to update on the event generated via the Simulation engine and not a background refreshing?
So if this is the case I think windows should be unregistering themselves as a listener when they are closed down. Also I don't think we should keep the UnitWindows in memory once closed. I know you have said previously that all the tabs in a window are created when the UnitWindow is displayed but I think we need to create the tabs on demand when displayed. Plus I'm now convinced we need a new ```BuildingWindow``` that contains details about a specific Building. The Building tab in the Settlement Window is massive.
I'm no expecting us to do anything about this now but to work out how we want it to behave for a future "project".

As to the Web UI, I'm doing a Responsive Design course over Christmas which should help change my hacking of Web UIs into some more usable. Then I want to create a pilot next year of a basic Web UI.

**Reply by mokun** [reply]
> I don't think we should keep the UnitWindows in memory once closed.

Ok. It takes up some memory but not much. It seems to me the main thing is the lack of organization in the desktop workspace. 

> we need to create the tabs on demand when displayed

Agree. Namely we only need to generate the panel for each building that the player is interested in knowing. 

Currently clicking on the building Tab would always generates all the functions panels for all buildings in a settlement at once and there is no real need to do so.

This could be just some simple changes we can make real quick.

**Reply by mokun** [reply]
> As to the Web UI, I'm doing a Responsive Design course over Christmas which should help change my hacking of Web UIs into some more usable. Then I want to create a pilot next year of a basic Web UI.

That's awesome ! 

Yes the future is that we tweak mars-sim to be able to host on a server as a persistent simulation.

**Reply by mokun** [reply]
> Then I want to create a pilot next year of a basic Web UI

Yes. There's so many extremely powerful animation we can use to dress up mars-sim. 

I just look at this one called [Framer Motion 3D](https://www.framer.com/docs/three-introduction/).

In this [CodeSandbox](https://codesandbox.io/s/framer-motion-3d-star-button-forked-1pm89?from-embed) site, we can even tweak and customize what it does.

**Comment by mokun** [comment]
Another thing is about standardization and customization. 

The thing is that currently we limit the dragging of a window in such a way that the whole width of it must fit the desktop. 

Initially I was limiting how the height of the unit window must fit inside the desktop but it doesn't work well after we allow resizing the desktop.

So we can work toward standardizing the location of where each panel should pop up. 

I supposed this is what most strategy or simulation games would prefer as it helps player focus on the mechanics of the games. 

But for real world mission system, it's not a bad thing to have more degree of UI customization.

**Comment by mokun** [comment]
> Web UI based on a REST api

btw, the next version Java, namely, Java 18, will include a simple [web server](https://openjdk.java.net/jeps/408) for testing and educational purpose.

Having said that, are you still interested in using springboot or switch to something else for bootstrapping ?

**Reply by bevans2000** [reply]
Yes SpringBoot offers a lot of supporting libraries and classes, e.g. OAuth, JSON, Swagger, REST controller, security

**Reply by bevans2000** [reply]
Relates to #531
